### ISSUE TITLE
[BUG] nan values in validation loss (and solution)

### ISSUE THREAD

**Issue by ChristiaanWewer** [neuralhydrology/neuralhydrology#258]
**Describe the bug**
Hi there, I cloned the latest version from github to train our models, however I found a bug. When training without any of the new fancy nan handling methods, but with nan's present in the training and validation data the validation loss becomes a nan value. This is obviously undesired.

**Solution**
I took a little dive in the code and in neuralhydrology/datasetzoo/basedataset.py at line 621 to 625 x_d_validate is only used to flag samples with nans during training so we can avoid them. However, in the validation phase we also want to flag them when we are not using any nan handling methods. Thus the solution would be to turn 
```
                if self.is_train:
                    x_d_validate = [np.concatenate([v for v in x_d[freq].values()], axis=-1)
                                    for freq in self.frequencies]
                else:
                    x_d_validate = None
```
into
```
                if self.is_train or not self.cfg.nan_handling_method:
                    x_d_validate = [np.concatenate([v for v in x_d[freq].values()], axis=-1)
                                    for freq in self.frequencies]
                else:
                    x_d_validate = None
```
I tested this with our own data and without any nan handling methods, which gives me a validation loss that is not nan. I did not test this with any of the fancy nan handling methods.

Let me know what you guys think!
Cheers,
Christiaan

**Comment by gauchm** [comment]
Hi Christiaan,
Thanks for reporting! That does sound suboptimal. I'm not sure I fully understand what's going on yet, but I have some thoughts:
- We don't discard samples without targets in validation and that is on purpose, because we want to make contiguous predictions so that you can plot a hydrograph of predictions. The solution you propose above would absolutely fix the issue you are facing, but it would break this point -- we'd no longer get contiguous predictions out of the validation/test periods.
- We do already filter NaNs when reporting metrics and losses in the tester (see [tester](https://github.com/neuralhydrology/neuralhydrology/blob/d1022cf63527096fc66b8ff2dfc765c33c4473b4/neuralhydrology/evaluation/tester.py#L487)). But: the losses that we average there are already averaged over each minibatch. So my best explanation for what's happening in your case is that every single minibatch has at least one missing target time step (so that each minibatch loss is NaN and therefore the overall nanmean is also NaN). Is that correct?

I guess maybe we could give the loss an argument to not reduce the per-sample losses during validation, so that we can do an overall nanmean instead of nanmean of means? Not sure I like this solution... I'd prefer to find something more simple than that.
